File: scripts/javascript/04_coastal_exposure.js
/**********************************************************************
 * 04_coastal_exposure.js   —  SEA LEVEL & LOW-LYING COASTAL EXPOSURE
 * SPREP / UNEP GEE Climate Workshop 2026
 *
 * GOAL: Highlight the low-elevation land most exposed to sea-level rise
 *       and storm surge, and where surface water already sits.
 *
 * DATASETS:
 *   - NASADEM elevation (NASA/NASADEM_HGT/001), 'elevation' (metres, ~30 m)
 *   - JRC Global Surface Water (JRC/GSW1_4/GlobalSurfaceWater), 'occurrence' %
 *
 * NOTE: 30 m DEMs are coarse for very flat atolls, so treat this as an
 *       awareness / screening tool. Best results on the larger high islands
 *       (Samoa, Fiji, Vanuatu, Solomon Islands, PNG).
 *
 * TO LOCALISE: change COUNTRY below.
 **********************************************************************/

// ===== 1. COUNTRY SELECTOR =========================================
var COUNTRY      = 'Samoa';
var FLOOD_HEIGHT = 5;    // metres above sea level to flag as "low-lying"

var LSIB_NAMES = {
  'Fiji':'Fiji', 'Papua New Guinea':'Papua New Guinea', 'Solomon Islands':'Solomon Is',
  'Vanuatu':'Vanuatu', 'Samoa':'Samoa', 'New Caledonia':'New Caledonia'};
var POINT_AOI = {
  'Tonga':[-174.8,-20.0,300000], 'Palau':[134.58,7.5,120000], 'Tuvalu':[178.5,-7.8,350000],
  'Kiribati':[173.0,1.4,500000], 'Nauru':[166.93,-0.52,40000], 'Niue':[-169.87,-19.05,40000],
  'Cook Islands':[-159.78,-21.23,300000], 'Marshall Islands':[169.0,8.0,600000],
  'Federated States of Micronesia':[158.21,6.92,500000], 'Tokelau':[-171.85,-9.2,150000],
  'American Samoa':[-170.70,-14.30,60000], 'French Polynesia':[-149.5,-17.6,200000],
  'Guam':[144.79,13.44,60000], 'Northern Mariana Islands':[145.6,15.6,200000],
  'Wallis & Futuna':[-176.2,-13.3,80000]};
var aoi, outline;
if (LSIB_NAMES[COUNTRY]) {
  outline = ee.FeatureCollection('USDOS/LSIB_SIMPLE/2017')
              .filter(ee.Filter.eq('country_na', LSIB_NAMES[COUNTRY]));
  aoi = outline.geometry();
} else {
  var pt = POINT_AOI[COUNTRY];
  aoi = ee.Geometry.Point([pt[0], pt[1]]).buffer(pt[2]);
  outline = ee.FeatureCollection([ee.Feature(aoi)]);
}
Map.centerObject(aoi, 10);

// ===== 2. ELEVATION & LOW-LYING ZONE ==============================
var dem = ee.Image('NASA/NASADEM_HGT/001').select('elevation').clip(aoi);
var elevPalette = ['d73027','fc8d59','fee08b','d9ef8b','91cf60','1a9850'];
Map.addLayer(dem, {min: 0, max: 200, palette: elevPalette}, 'Elevation (m)');

var lowLying = dem.lte(FLOOD_HEIGHT).selfMask();   // ocean is masked (no data)
Map.addLayer(lowLying, {palette: ['ff0000']},
             'Land below ' + FLOOD_HEIGHT + ' m (most exposed)');

// ===== 3. EXISTING SURFACE WATER ==================================
var water = ee.Image('JRC/GSW1_4/GlobalSurfaceWater').select('occurrence').clip(aoi);
Map.addLayer(water.updateMask(water.gt(20)),
             {min: 0, max: 100, palette: ['c7e9b4','41b6c4','225ea8','0c2c84']},
             'Surface water occurrence 1984-2021 (%)');

// ===== 4. HOW MUCH LAND IS EXPOSED? (area in km²) ==================
var pixelAreaKm2 = ee.Image.pixelArea().divide(1e6);
print('Land area below ' + FLOOD_HEIGHT + ' m in ' + COUNTRY + ' (km²):',
  pixelAreaKm2.updateMask(lowLying).reduceRegion({
    reducer: ee.Reducer.sum(), geometry: aoi, scale: 30, maxPixels: 1e13}));
print('Total land area of ' + COUNTRY + ' (km²):',
  pixelAreaKm2.updateMask(dem.gte(-100)).reduceRegion({
    reducer: ee.Reducer.sum(), geometry: aoi, scale: 30, maxPixels: 1e13}));

Map.addLayer(outline.style({color: 'black', fillColor: '00000000', width: 1}), {}, COUNTRY + ' outline');

// ===== 5. EXPORT the low-lying zone (optional) =====================
Export.image.toDrive({
  image: lowLying.unmask(0),
  description: COUNTRY.replace(/[ ,]/g, '_') + '_LowLying_below' + FLOOD_HEIGHT + 'm',
  folder: 'GEE_Workshop_2026',
  fileNamePrefix: COUNTRY.replace(/[ ,]/g, '_') + '_lowlying_' + FLOOD_HEIGHT + 'm',
  scale: 30, region: aoi.bounds(), maxPixels: 1e13
});

File: scripts/python/04_coastal_exposure.py
"""
04_coastal_exposure.py  —  SEA LEVEL & LOW-LYING COASTAL EXPOSURE (Python / geemap)
SPREP / UNEP GEE Climate Workshop 2026

Python equivalent of javascript/04_coastal_exposure.js.
Flags low-lying land exposed to sea-level rise / storm surge, shows where
surface water already sits, and prints the exposed land area in km².

Datasets:
  NASADEM elevation (NASA/NASADEM_HGT/001) band 'elevation' (metres)
  JRC Global Surface Water (JRC/GSW1_4/GlobalSurfaceWater) band 'occurrence' (%)

NOTE: 30 m DEMs are coarse for very flat atolls — treat as a screening /
awareness tool, not a survey-grade flood model. Use a larger high island
(Samoa, Fiji, Tonga) for this exercise.
"""

import ee
import geemap
# Shared country selector (the 14 SPREP member countries). In Colab / Jupyter
# this helper file may not be present, so fetch it from the workshop repo if so.
try:
    from _pacific_aoi import get_country, get_outline
except ModuleNotFoundError:
    import urllib.request
    urllib.request.urlretrieve(
        'https://raw.githubusercontent.com/tavitas/sprep-gee-climate-workshop/'
        'main/scripts/python/_pacific_aoi.py', '_pacific_aoi.py')
    from _pacific_aoi import get_country, get_outline

ee.Initialize(project='your-project-id')   # <-- your registered project

# ===== 1. SETTINGS =====
COUNTRY      = 'Samoa'   # best on larger high islands (Fiji, Vanuatu, Solomon Islands)
FLOOD_HEIGHT = 5    # metres above sea level flagged as "low-lying"

aoi = get_country(COUNTRY)       # ee.Geometry
outline = get_outline(COUNTRY)   # ee.FeatureCollection for drawing

# ===== 2. ELEVATION & LOW-LYING ZONE =====
dem = ee.Image('NASA/NASADEM_HGT/001').select('elevation').clip(aoi)
elev_palette = ['d73027', 'fc8d59', 'fee08b', 'd9ef8b', '91cf60', '1a9850']

Map = geemap.Map()
Map.centerObject(aoi, 10)
Map.addLayer(dem, {'min': 0, 'max': 200, 'palette': elev_palette}, 'Elevation (m)')

low_lying = dem.lte(FLOOD_HEIGHT).selfMask()
Map.addLayer(low_lying, {'palette': ['ff0000']},
             f'Land below {FLOOD_HEIGHT} m (most exposed)')

# ===== 3. EXISTING SURFACE WATER =====
water = ee.Image('JRC/GSW1_4/GlobalSurfaceWater').select('occurrence').clip(aoi)
Map.addLayer(water.updateMask(water.gt(20)),
             {'min': 0, 'max': 100, 'palette': ['c7e9b4', '41b6c4', '225ea8', '0c2c84']},
             'Surface water occurrence 1984-2021 (%)')
Map.addLayer(outline.style(color='black', fillColor='00000000', width=1), {}, f'{COUNTRY} outline')
Map   # displays in a notebook

# ===== 4. HOW MUCH LAND IS EXPOSED? (area in km²) =====
km2 = ee.Image.pixelArea().divide(1e6)

exposed = km2.updateMask(low_lying).reduceRegion(
    reducer=ee.Reducer.sum(), geometry=aoi, scale=30, maxPixels=int(1e13)
).get('area').getInfo()

total = km2.updateMask(dem.gte(-100)).reduceRegion(
    reducer=ee.Reducer.sum(), geometry=aoi, scale=30, maxPixels=int(1e13)
).get('area').getInfo()

print(f'{COUNTRY}: land below {FLOOD_HEIGHT} m = {exposed:,.1f} km²')
print(f'{COUNTRY}: total land area      = {total:,.1f} km²')
if total:
    print(f'Share of land below {FLOOD_HEIGHT} m = {100 * exposed / total:.1f}%')

# ===== 5. EXPORT the low-lying zone (optional) =====
task = ee.batch.Export.image.toDrive(
    image=low_lying.unmask(0),
    description=f'{COUNTRY}_LowLying_below{FLOOD_HEIGHT}m',
    folder='GEE_Workshop_2026',
    fileNamePrefix=f'{COUNTRY.replace(" ", "_")}_lowlying_{FLOOD_HEIGHT}m',
    scale=30,
    region=aoi.bounds(),
    maxPixels=int(1e13))
task.start()
print('Export started — check the Tasks panel at https://code.earthengine.google.com')
